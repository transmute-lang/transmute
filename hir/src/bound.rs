use std::fmt::{Debug, Formatter};
use transmute_core::ids::SymbolId;

pub trait BoundState {}

#[derive(Debug, Clone, PartialEq)]
pub struct Unbound;

impl BoundState for Unbound {}

#[derive(Clone, PartialEq)]
pub struct Bound(pub SymbolId);

impl BoundState for Bound {}

impl Debug for Bound {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "Bound({})", self.0)
    }
}

#[derive(PartialEq)]
pub struct BoundMultiple(pub Vec<SymbolId>);

impl BoundState for BoundMultiple {}

impl Debug for BoundMultiple {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "Bound({:?})", self.0)
    }
}
