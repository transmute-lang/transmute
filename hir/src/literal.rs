use std::fmt::{Debug, Formatter};
use transmute_ast::literal::Literal as AstLiteral;
use transmute_ast::literal::LiteralKind as AstLiteralKind;
use transmute_core::ids::IdentRefId;
use transmute_core::span::Span;

#[derive(Debug, Clone, PartialEq)]
pub struct Literal {
    pub kind: LiteralKind,
    pub span: Span,
}

impl Literal {
    pub fn new(kind: LiteralKind, span: Span) -> Self {
        Self { kind, span }
    }
}

impl From<AstLiteral> for Literal {
    fn from(value: AstLiteral) -> Self {
        Self {
            span: value.span.clone(),
            kind: LiteralKind::from(value.kind),
        }
    }
}

#[derive(Clone, PartialEq)]
pub enum LiteralKind {
    Boolean(bool),
    Identifier(IdentRefId),
    Number(i64),
}

impl From<AstLiteralKind> for LiteralKind {
    fn from(value: AstLiteralKind) -> Self {
        match value {
            AstLiteralKind::Boolean(b) => Self::Boolean(b),
            AstLiteralKind::Identifier(i) => Self::Identifier(i),
            AstLiteralKind::Number(n) => Self::Number(n),
        }
    }
}

impl Debug for LiteralKind {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            LiteralKind::Boolean(v) => write!(f, "Boolean({v})"),
            LiteralKind::Identifier(v) => write!(f, "IdentRefId({v})"),
            LiteralKind::Number(v) => write!(f, "Number({v})"),
        }
    }
}
