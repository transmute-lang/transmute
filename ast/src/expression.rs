use crate::literal::Literal;
use crate::operators::{BinaryOperator, UnaryOperator};
use std::fmt::Debug;
use transmute_core::ids::{ExprId, IdentRefId, StmtId};
use transmute_core::span::Span;

#[derive(Debug, Clone, PartialEq)]
pub struct Expression {
    pub id: ExprId,
    pub kind: ExpressionKind,
    pub span: Span,
}

impl Expression {
    pub fn new(id: ExprId, kind: ExpressionKind, span: Span) -> Self {
        Self { id, kind, span }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum ExpressionKind {
    Assignment(Target, ExprId),
    If(ExprId, ExprId, Option<ExprId>),
    Literal(Literal),
    Binary(ExprId, BinaryOperator, ExprId),
    Unary(UnaryOperator, ExprId),
    Access(ExprId, IdentRefId),
    FunctionCall(IdentRefId, Vec<ExprId>),
    While(ExprId, ExprId),
    Block(Vec<StmtId>),
    StructInstantiation(IdentRefId, Vec<(IdentRefId, ExprId)>),
    /// A dummy expression kind, inserted by the parser when the expression could not be parsed
    Dummy,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Target {
    Direct(IdentRefId),
    // The expression id is of kind ExpressionKind::Access
    Indirect(ExprId),
}
