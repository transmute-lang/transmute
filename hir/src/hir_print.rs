use crate::bound::Bound;
use crate::expression::{Expression, ExpressionKind, Target};
use crate::literal::{Literal, LiteralKind};
use crate::natives::Type;
use crate::statement::{Implementation, Statement, StatementKind, TypeDefKind};
use crate::typed::Typed;
use crate::Hir;
use bimap::BiHashMap;
use std::fmt::Write;
use transmute_core::id;
use transmute_core::ids::{ExprId, IdentId, IdentRefId, StmtId, SymbolId, TypeDefId, TypeId};

pub trait HirPrint {
    fn hir_print<W>(
        &self,
        ctx: &mut HirPrintContext<'_>,
        opts: &Options,
        f: &mut W,
    ) -> std::fmt::Result
    where
        W: Write;
}

#[derive(Default)]
pub struct Options {}

impl HirPrint for Literal {
    fn hir_print<W>(
        &self,
        ctx: &mut HirPrintContext<'_>,
        _opts: &Options,
        f: &mut W,
    ) -> std::fmt::Result
    where
        W: Write,
    {
        match &self.kind {
            LiteralKind::Boolean(b) => {
                writeln!(
                    f,
                    "{indent}Bool: {b}",
                    indent = ctx.indent_type_id(ctx.resolve_type_id_by_type(Type::Boolean))
                )
            }
            LiteralKind::Identifier(ident) => {
                writeln!(
                    f,
                    "{indent}Ident: {ident}",
                    indent = ctx.indent_symbol_ref(ctx.identifier_ref_symbol(*ident)),
                    ident = ctx.identifier_ref_name(*ident),
                )
            }
            LiteralKind::Number(n) => {
                writeln!(
                    f,
                    "{indent}Number: {n}",
                    indent = ctx.indent_type_id(ctx.resolve_type_id_by_type(Type::Number))
                )
            }
            LiteralKind::String(n) => {
                // todo type
                writeln!(f, "{indent}String: {n}", indent = ctx.indent())
            }
        }
    }
}

impl HirPrint for Expression<Typed> {
    fn hir_print<W>(
        &self,
        ctx: &mut HirPrintContext<'_>,
        opts: &Options,
        f: &mut W,
    ) -> std::fmt::Result
    where
        W: Write,
    {
        if let Some(tag) = ctx.tag.take() {
            writeln!(
                f,
                "{indent}{tag}:expr {id}",
                indent = ctx.indent_type_id(self.resolved_type_id()),
                id = id!(self.id)
            )?;
        } else {
            writeln!(
                f,
                "{indent}expr {id}",
                indent = ctx.indent_type_id(self.resolved_type_id()),
                id = id!(self.id)
            )?;
        }

        ctx.next_level();
        match &self.kind {
            ExpressionKind::Assignment(Target::Direct(target), expr_id) => {
                writeln!(
                    f,
                    "{indent}Assignment target={ident}",
                    indent = ctx.indent_symbol_ref(ctx.identifier_ref_symbol(*target)),
                    ident = ctx.identifier_ref_name(*target)
                )?;
                ctx.last = true;
                ctx.next_level();
                ctx.last = true;
                ctx.hir_print_expression(*expr_id, opts, f)?;
                ctx.prev_level();
            }
            ExpressionKind::Assignment(Target::Indirect(lhs_expr_id), rhs_expr_id) => {
                writeln!(
                    f,
                    "{indent}Assignment",
                    indent = ctx.indent_type_id(ctx.hir.expressions[self.id].resolved_type_id())
                )?;
                ctx.last = true;
                ctx.next_level();

                ctx.tag = Some("target");
                ctx.hir_print_expression(*lhs_expr_id, opts, f)?;
                ctx.tag = Some("value");
                ctx.last = true;
                ctx.hir_print_expression(*rhs_expr_id, opts, f)?;
                ctx.prev_level();
            }
            ExpressionKind::If(cond_id, true_id, false_id) => {
                writeln!(
                    f,
                    "{indent}If",
                    indent = ctx.indent_type_id(ctx.hir.expressions[self.id].resolved_type_id())
                )?;
                ctx.last = true;
                ctx.next_level();

                ctx.tag = Some("cond");
                ctx.hir_print_expression(*cond_id, opts, f)?;

                ctx.tag = Some("true");
                ctx.last = !false_id.is_some();
                ctx.hir_print_expression(*true_id, opts, f)?;

                if let Some(false_id) = false_id {
                    ctx.tag = Some("false");
                    ctx.last = true;
                    ctx.hir_print_expression(*false_id, opts, f)?;
                }

                ctx.prev_level();
            }
            ExpressionKind::Literal(lit) => lit.hir_print(ctx, opts, f)?,
            ExpressionKind::Access(expr_id, ident_ref_id) => {
                writeln!(
                    f,
                    "{indent}Access",
                    indent = ctx.indent_symbol_ref(ctx.identifier_ref_symbol(*ident_ref_id))
                )?;

                ctx.last = true;
                ctx.next_level();
                ctx.hir_print_expression(*expr_id, opts, f)?;
                writeln!(
                    f,
                    "{indent}ident={ident}",
                    indent = ctx.indent_symbol_ref(ctx.identifier_ref_symbol(*ident_ref_id)),
                    ident = ctx.identifier_ref_name(*ident_ref_id)
                )?;
                ctx.prev_level();
            }
            ExpressionKind::FunctionCall(expr_id, param_ids) => {
                writeln!(
                    f,
                    "{indent}Call",
                    indent = ctx.indent_type_id(ctx.hir.expressions[self.id].resolved_type_id())
                )?;

                ctx.last = true;
                ctx.next_level();
                ctx.hir_print_expression(*expr_id, opts, f)?;

                if param_ids.is_empty() {
                    writeln!(f, "{indent}params=[]", indent = ctx.indent())?;
                } else {
                    writeln!(f, "{indent}params", indent = ctx.indent())?;
                    ctx.last = true;
                    ctx.next_level();
                    for (i, param) in param_ids.iter().enumerate() {
                        ctx.last = i + 1 == param_ids.len();
                        ctx.hir_print_expression(*param, opts, f)?;
                    }
                    ctx.prev_level();
                }

                ctx.prev_level();
            }
            ExpressionKind::While(cond_id, expr_id) => {
                writeln!(
                    f,
                    "{indent}While",
                    indent = ctx.indent_type_id(ctx.hir.expressions[self.id].resolved_type_id())
                )?;
                ctx.last = true;
                ctx.next_level();

                ctx.tag = Some("cond");
                ctx.hir_print_expression(*cond_id, opts, f)?;
                ctx.tag = Some("body");
                ctx.last = true;
                ctx.hir_print_expression(*expr_id, opts, f)?;

                ctx.prev_level();
            }
            ExpressionKind::Block(stmt_ids) => {
                if stmt_ids.is_empty() {
                    writeln!(
                        f,
                        "{indent}Block=[]",
                        indent =
                            ctx.indent_type_id(ctx.hir.expressions[self.id].resolved_type_id())
                    )?;
                } else {
                    writeln!(
                        f,
                        "{indent}Block",
                        indent =
                            ctx.indent_type_id(ctx.hir.expressions[self.id].resolved_type_id())
                    )?;
                    ctx.last = true;
                    ctx.next_level();
                    for (i, stmt_id) in stmt_ids.iter().enumerate() {
                        ctx.last = i + 1 == stmt_ids.len();
                        ctx.hir_print_statement(*stmt_id, opts, f)?;
                    }
                    ctx.prev_level();
                }
            }
            ExpressionKind::StructInstantiation(ident_ref_id, fields) => {
                writeln!(
                    f,
                    "{indent}StructInstantiate name={ident}",
                    indent = ctx.indent_symbol_ref(ctx.identifier_ref_symbol(*ident_ref_id)),
                    ident = ctx.identifier_ref_name(*ident_ref_id)
                )?;
                ctx.last = true;
                ctx.next_level();
                for (i, (field_ident, field_value)) in fields.iter().enumerate() {
                    writeln!(
                        f,
                        "{indent}field name={ident}",
                        indent = ctx.indent_symbol_ref(ctx.identifier_ref_symbol(*field_ident)),
                        ident = ctx.identifier_ref_name(*field_ident),
                    )?;
                    ctx.last = i + 1 == fields.len();
                    ctx.next_level();
                    ctx.last = true;
                    ctx.hir_print_expression(*field_value, opts, f)?;
                    ctx.prev_level();
                }
                ctx.prev_level();
            }
            ExpressionKind::ArrayInstantiation(values) => {
                writeln!(
                    f,
                    "{indent}ArrayInstantiate",
                    indent = ctx.indent_type_id(ctx.hir.expressions[self.id].resolved_type_id()),
                )?;
                ctx.last = true;
                ctx.next_level();
                for (i, expr_id) in values.iter().enumerate() {
                    ctx.last = i + 1 == values.len();
                    ctx.hir_print_expression(*expr_id, opts, f)?;
                }
                ctx.prev_level();
            }
            ExpressionKind::ArrayAccess(base_expr_id, index_expr_id) => {
                writeln!(
                    f,
                    "{indent}ArrayAccess",
                    indent = ctx.indent_type_id(ctx.hir.expressions[self.id].resolved_type_id()),
                )?;

                ctx.last = true;
                ctx.next_level();
                ctx.hir_print_expression(*base_expr_id, opts, f)?;
                ctx.last = true;
                ctx.hir_print_expression(*index_expr_id, opts, f)?;
                ctx.prev_level();
            }
        }
        ctx.prev_level();
        Ok(())
    }
}

impl HirPrint for Statement<Typed, Bound> {
    fn hir_print<W>(
        &self,
        ctx: &mut HirPrintContext<'_>,
        opts: &Options,
        f: &mut W,
    ) -> std::fmt::Result
    where
        W: Write,
    {
        if let Some(tag) = ctx.tag.take() {
            writeln!(
                f,
                "{indent}{tag}:stmt {id}",
                indent = ctx.indent(),
                id = id!(self.id)
            )?;
        } else {
            writeln!(
                f,
                "{indent}stmt {id}",
                indent = ctx.indent(),
                id = id!(self.id)
            )?;
        }
        ctx.next_level();
        match &self.kind {
            StatementKind::Expression(expr_id) => {
                writeln!(
                    f,
                    "{indent}Expression",
                    indent = ctx.indent_type_id(ctx.hir.expressions[*expr_id].resolved_type_id()),
                )?;
                ctx.last = true;
                ctx.next_level();
                ctx.last = true;
                ctx.hir_print_expression(*expr_id, opts, f)?;
                ctx.prev_level();
            }
            StatementKind::Namespace(ident, _, stmts) => {
                writeln!(
                    f,
                    "{indent}Namespace name={ident}",
                    indent = ctx.indent_symbol_def(ident.resolved_symbol_id()),
                    ident = ctx.identifier(ident.id)
                )?;

                ctx.last = true;
                ctx.next_level();
                if stmts.is_empty() {
                    writeln!(f, "{indent}stmts=[]", indent = ctx.indent(),)?;
                } else {
                    for (i, stmt) in stmts.iter().enumerate() {
                        ctx.last = i + 1 == stmts.len();
                        ctx.hir_print_statement(*stmt, opts, f)?;
                    }
                }
                ctx.prev_level();
            }
            StatementKind::Let(ident, expr_id) => {
                writeln!(
                    f,
                    "{indent}Let name={ident}",
                    indent = ctx.indent_symbol_def(ident.resolved_symbol_id()),
                    ident = ctx.identifier(ident.id)
                )?;
                ctx.last = true;
                ctx.next_level();
                ctx.last = true;
                ctx.hir_print_expression(*expr_id, opts, f)?;
                ctx.prev_level();
            }
            StatementKind::Ret(Some(expr_id), _mode) => {
                writeln!(
                    f,
                    "{indent}Ret",
                    indent = ctx.indent_type_id(ctx.hir.expressions[*expr_id].resolved_type_id())
                )?;
                ctx.last = true;
                ctx.next_level();
                ctx.last = true;
                ctx.hir_print_expression(*expr_id, opts, f)?;
                ctx.prev_level();
            }
            StatementKind::Ret(None, mode) => {
                writeln!(
                    f,
                    "{indent}Ret ({mode})",
                    indent = ctx.indent_type_id(ctx.resolve_type_id_by_type(Type::Void)),
                    mode = mode.as_str()
                )?;
            }
            StatementKind::LetFn(ident, annotations, parameters, ret_type, expr_id) => {
                writeln!(
                    f,
                    "{indent}Fn name={ident}{native}",
                    indent = ctx.indent_symbol_def(ident.resolved_symbol_id()),
                    ident = ctx.identifier(ident.id),
                    native = match expr_id {
                        Implementation::Provided(_) => "",
                        _ => " (native)",
                    }
                )?;
                ctx.last = true;
                ctx.next_level();

                if annotations.is_empty() {
                    writeln!(f, "{indent}annotations=[]", indent = ctx.indent())?;
                } else {
                    writeln!(f, "{indent}annotations", indent = ctx.indent())?;
                    ctx.next_level();
                    for annotation in annotations {
                        writeln!(
                            f,
                            "{indent}{ident}",
                            indent =
                                ctx.indent_symbol_ref(ctx.identifier_ref_symbol(
                                    *annotation.ident_ref_ids.last().unwrap()
                                )),
                            ident = annotation
                                .ident_ref_ids
                                .iter()
                                .map(|ident_ref_id| ctx.identifier_ref_name(*ident_ref_id))
                                .collect::<Vec<_>>()
                                .join("."),
                        )?;
                    }
                    ctx.prev_level();
                }

                if parameters.is_empty() {
                    writeln!(f, "{indent}params=[]", indent = ctx.indent())?;
                } else {
                    writeln!(f, "{indent}params", indent = ctx.indent())?;
                    ctx.next_level();
                    for parameter in parameters.iter() {
                        write!(
                            f,
                            "{indent}{ident}: ",
                            indent = ctx.indent_symbol_def(parameter.resolved_symbol_id()),
                            ident = ctx.identifier(parameter.identifier.id),
                        )?;
                        parameter.type_def_id.hir_print(ctx, opts, f)?;
                        writeln!(f)?;
                    }
                    ctx.prev_level();
                }

                if let Some(ret_type) = &ret_type.type_def_id {
                    write!(
                        f,
                        "{indent}ret=",
                        indent = ctx.indent_type_id(ctx.resolve_type_id_by_type_def_id(*ret_type))
                    )?;
                    ret_type.hir_print(ctx, opts, f)?;
                    writeln!(f)?;
                } else {
                    writeln!(
                        f,
                        "{indent}ret=[]",
                        indent = ctx.indent_type_id(ctx.resolve_type_id_by_type(Type::Void))
                    )?;
                }

                if let Implementation::Provided(expr_id) = expr_id {
                    ctx.tag = Some("body");
                    ctx.last = true;
                    ctx.hir_print_expression(*expr_id, opts, f)?;
                }

                ctx.prev_level();
            }
            StatementKind::Struct(ident, _annotations, fields) => {
                // todo use annotations
                writeln!(
                    f,
                    "{indent}Struct name={ident}{native}",
                    indent = ctx.indent_symbol_def(ident.resolved_symbol_id()),
                    ident = ctx.identifier(ident.id),
                    native = match fields {
                        Implementation::Provided(_) => "",
                        _ => " (native)",
                    }
                )?;
                ctx.last = true;
                ctx.next_level();

                if let Implementation::Provided(fields) = fields {
                    if fields.is_empty() {
                        writeln!(f, "{indent}fields=[]", indent = ctx.indent())?;
                    } else {
                        for field in fields.iter() {
                            write!(
                                f,
                                "{indent}{ident}: ",
                                indent = ctx.indent_symbol_def(field.resolved_symbol_id()),
                                ident = ctx.identifier(field.identifier.id),
                            )?;
                            field.type_def_id.hir_print(ctx, opts, f)?;
                            writeln!(f)?;
                        }
                    }
                }
                ctx.prev_level();
            }
            StatementKind::Use(_) => {
                unreachable!("for as long as the resolver removes the use statements")
            }
            StatementKind::Annotation(ident) => {
                writeln!(
                    f,
                    "{indent}Annotation name={ident}",
                    indent = ctx.indent_symbol_def(ident.resolved_symbol_id()),
                    ident = ctx.identifier(ident.id)
                )?;
            }
        }
        ctx.prev_level();
        Ok(())
    }
}

impl HirPrint for TypeDefId {
    fn hir_print<W>(
        &self,
        ctx: &mut HirPrintContext<'_>,
        opts: &Options,
        f: &mut W,
    ) -> std::fmt::Result
    where
        W: Write,
    {
        match &ctx.hir.type_defs[*self].kind {
            TypeDefKind::Simple(ident_ref_ids) => {
                write!(
                    f,
                    "{}",
                    ident_ref_ids
                        .iter()
                        .map(|ident_ref| ctx.identifier_ref_name(*ident_ref))
                        .collect::<Vec<_>>()
                        .join(".")
                )
            }
            TypeDefKind::Array(base, len) => {
                write!(f, "[")?;
                base.hir_print(ctx, opts, f)?;
                write!(f, "; {len}]")
            }
        }
    }
}

impl Hir<Typed, Bound> {
    pub fn hir_print<W>(&self, opts: &Options, f: &mut W) -> std::fmt::Result
    where
        W: Write,
    {
        let types = self.types.iter().collect::<BiHashMap<_, _>>();

        let mut ctx = HirPrintContext {
            hir: self,
            types,
            indent: String::with_capacity(3 * 32),
            tag: None,
            last: true,
        };
        ctx.hir_print_statement(self.root, opts, f)?;
        writeln!(f, "\n--- types ---")?;
        for (id, ty) in &ctx.hir.types {
            write!(f, "{id:-5}: ")?;
            self.print_type(f, ty)?;
            writeln!(f,)?;
        }
        Ok(())
    }

    fn print_type<W>(&self, f: &mut W, ty: &Type) -> std::fmt::Result
    where
        W: Write,
    {
        match ty {
            Type::Invalid => write!(f, "invalid"),
            Type::Boolean => write!(f, "boolean"),
            Type::Number => write!(f, "number"),
            Type::Function(params, ret) => {
                write!(f, "f(")?;
                for (i, param) in params.iter().enumerate() {
                    if i > 0 {
                        write!(f, ", ")?;
                    }
                    write!(f, "{}", id!(*param))?;
                }
                write!(f, ") : {}", id!(*ret))
            }
            Type::Struct(stmt_id) => {
                let symbol_id = match &self.statements[*stmt_id].kind {
                    StatementKind::Struct(ident, _, _) => ident.resolved_symbol_id(),
                    _ => panic!("struct expected"),
                };
                write!(f, "struct {}", id!(symbol_id))
            }
            Type::Array(type_id, size) => {
                write!(f, "[{type_id}; {size}]", type_id = id!(*type_id))
            }
            Type::Void => write!(f, "void"),
            Type::None => write!(f, "none"),
        }
    }
}

pub struct HirPrintContext<'a> {
    hir: &'a Hir<Typed, Bound>,
    types: BiHashMap<TypeId, &'a Type>,
    indent: String,
    tag: Option<&'static str>,
    last: bool,
}

const NODE: &str = "+- ";
const INDENT: &str = "|  ";

// const NODE: &str = "";
// const INDENT: &str = "  ";

impl HirPrintContext<'_> {
    fn indent(&self) -> String {
        format!("               {}{NODE}", self.indent)
    }

    fn indent_symbol_def(&self, symbol_id: SymbolId) -> String {
        format!(
            "={symbol_id:-5}: {type_id:-5}  {}{NODE}",
            self.indent,
            symbol_id = id!(symbol_id),
            type_id = id!(self.hir.symbols[symbol_id].type_id),
        )
    }

    fn indent_symbol_ref(&self, symbol_id: SymbolId) -> String {
        format!(
            "@{symbol_id:-5}: {type_id:-5}  {}{NODE}",
            self.indent,
            symbol_id = id!(symbol_id),
            type_id = id!(self.hir.symbols[symbol_id].type_id),
        )
    }

    fn resolve_type_id_by_type_def_id(&self, type_def_id: TypeDefId) -> TypeId {
        match &self.hir.type_defs[type_def_id].kind {
            TypeDefKind::Simple(ident_ref_ids) => {
                self.hir.symbols[self.identifier_ref_symbol(*ident_ref_ids.last().unwrap())].type_id
            }
            TypeDefKind::Array(type_def_id_inner, size) => {
                let ty = Type::Array(
                    self.resolve_type_id_by_type_def_id(*type_def_id_inner),
                    *size,
                );
                self.resolve_type_id_by_type(ty)
            }
        }
    }

    fn resolve_type_id_by_type(&self, ty: Type) -> TypeId {
        *self.types.get_by_right(&ty).unwrap()
    }

    fn indent_type_id(&self, type_id: TypeId) -> String {
        format!(
            "@     : {type_id:-5}  {}{NODE}",
            self.indent,
            type_id = id!(type_id),
        )
    }

    fn next_level(&mut self) {
        if self.last {
            self.indent.push_str(&" ".repeat(INDENT.len()));
        } else {
            self.indent.push_str(INDENT);
        }
        self.last = false;
    }

    fn prev_level(&mut self) {
        for _ in 0..INDENT.len() {
            self.indent.pop();
        }
    }

    fn identifier(&self, ident_id: IdentId) -> &str {
        &self.hir.identifiers[ident_id]
    }

    fn identifier_ref_name(&self, ident_ref_id: IdentRefId) -> &str {
        self.identifier(self.hir.identifier_refs[ident_ref_id].ident.id)
    }

    fn identifier_ref_symbol(&self, ident_ref_id: IdentRefId) -> SymbolId {
        self.hir.identifier_refs[ident_ref_id].resolved_symbol_id()
    }

    fn hir_print_expression<W>(
        &mut self,
        expr_id: ExprId,
        opts: &Options,
        f: &mut W,
    ) -> std::fmt::Result
    where
        W: Write,
    {
        self.hir.expressions[expr_id].hir_print(self, opts, f)
    }

    fn hir_print_statement<W>(
        &mut self,
        stmt_id: StmtId,
        opts: &Options,
        f: &mut W,
    ) -> std::fmt::Result
    where
        W: Write,
    {
        self.hir.statements[stmt_id].hir_print(self, opts, f)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::Natives;
    use crate::UnresolvedHir;
    use insta::assert_snapshot;
    use std::env;
    use std::path::PathBuf;
    use transmute_ast::parse;
    use transmute_core::input::Input;

    macro_rules! t {
        ($name:ident) => {
            #[test]
            fn $name() {
                // fn f() {
                let stdlib_path = PathBuf::from(env::var("CARGO_MANIFEST_DIR").unwrap())
                    .parent()
                    .unwrap()
                    .join("stdlib")
                    .join("src")
                    .join("stdlib.tm");

                let inputs = vec![
                    Input::core(),
                    Input::from((
                        "test",
                        include_str!(concat!("../../examples/", stringify!($name), ".tm")),
                    )),
                    Input::try_from(stdlib_path).unwrap(),
                ];

                let ast = parse(inputs).1.unwrap();
                let hir = UnresolvedHir::from(ast).resolve(Natives::new()).unwrap();

                let mut s = String::new();
                hir.hir_print(&Options::default(), &mut s)
                    .expect("can print");
                assert_snapshot!(s);
            }
        };
    }

    t!(arrays_bounds);
    t!(arrays_if);
    t!(arrays_nested);
    t!(arrays_of_structs);
    t!(arrays_simple);
    t!(fibo_iter);
    t!(fibo_rec);
    t!(gc);
    t!(inner_function);
    t!(perimeter);
    t!(perimeter_lit_struct);
    t!(perimeter_var_struct);
    t!(print);
    t!(sibling_function);
    t!(string);
    t!(structs_anon);
    t!(structs_fn);
    t!(structs_inner);
    t!(structs_nested);
    t!(structs_of_arrays);
    t!(structs_simple);
    t!(void_fn);
}
