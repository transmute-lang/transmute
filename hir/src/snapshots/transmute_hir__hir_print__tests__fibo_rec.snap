---
source: hir/src/hir_print.rs
expression: s
---
               +- stmt 0
=    1:     1     +- Namespace name=<root>
                     +- stmt 1
=    2:     1        |  +- Namespace name=core
                     |     +- stmts=[]
                     +- stmt 9
=   34:    17        |  +- Fn name=main
                     |     +- annotations=[]
                     |     +- params=[]
@     :     1        |     +- ret=[]
@     :     2        |     +- body:expr 13
@     :     2        |        +- Block
                     |           +- stmt 6
=   36:     4        |           |  +- Let name=n
@     :     4        |           |     +- expr 6
@     :     4        |           |        +- Call
@     :    15        |           |           +- expr 0
@   32:    15        |           |           |  +- Ident: number_parse
                     |           |           +- params
@     :     6        |           |              +- expr 5
@     :     6        |           |                 +- Call
@     :    13        |           |                    +- expr 1
@   30:    13        |           |                    |  +- Ident: list_get
                     |           |                    +- params
@     :     5        |           |                       +- expr 3
@     :     5        |           |                       |  +- Call
@     :    12        |           |                       |     +- expr 2
@   29:    12        |           |                       |     |  +- Ident: args
                     |           |                       |     +- params=[]
@     :     4        |           |                       +- expr 4
@     :     4        |           |                          +- Number: 1
                     |           +- stmt 7
=   37:     4        |           |  +- Let name=f
@     :     4        |           |     +- expr 9
@     :     4        |           |        +- Call
@     :    10        |           |           +- expr 7
@   35:    10        |           |           |  +- Ident: fibo
                     |           |           +- params
@     :     4        |           |              +- expr 8
@   36:     4        |           |                 +- Ident: n
                     |           +- stmt 8
@     :     1        |              +- Ret
@     :     1        |                 +- expr 12
@     :     1        |                    +- Call
@     :    14        |                       +- expr 10
@   31:    14        |                       |  +- Ident: print
                     |                       +- params
@     :     4        |                          +- expr 11
@   37:     4        |                             +- Ident: f
                     +- stmt 13
=   35:    10        |  +- Fn name=fibo
                     |     +- annotations=[]
                     |     +- params
=   38:     4        |     |  +- n: number
@     :     2        |     +- ret=number
@     :     2        |     +- body:expr 31
@     :     2        |        +- Block
                     |           +- stmt 11
@     :     1        |           |  +- Expression
@     :     1        |           |     +- expr 19
@     :     1        |           |        +- If
@     :     3        |           |           +- cond:expr 16
@     :     3        |           |           |  +- Call
@     :     8        |           |           |     +- expr 38
@   20:     8        |           |           |     |  +- Ident: lt
                     |           |           |     +- params
@     :     4        |           |           |        +- expr 14
@   38:     4        |           |           |        |  +- Ident: n
@     :     4        |           |           |        +- expr 15
@     :     4        |           |           |           +- Number: 2
@     :     2        |           |           +- true:expr 18
@     :     2        |           |              +- Block
                     |           |                 +- stmt 10
@     :     4        |           |                    +- Ret
@     :     4        |           |                       +- expr 17
@   38:     4        |           |                          +- Ident: n
                     |           +- stmt 12
@     :     4        |              +- Ret
@     :     4        |                 +- expr 30
@     :     4        |                    +- Call
@     :     7        |                       +- expr 41
@   27:     7        |                       |  +- Ident: add
                     |                       +- params
@     :     4        |                          +- expr 24
@     :     4        |                          |  +- Call
@     :    10        |                          |     +- expr 20
@   35:    10        |                          |     |  +- Ident: fibo
                     |                          |     +- params
@     :     4        |                          |        +- expr 23
@     :     4        |                          |           +- Call
@     :     7        |                          |              +- expr 39
@   15:     7        |                          |              |  +- Ident: sub
                     |                          |              +- params
@     :     4        |                          |                 +- expr 21
@   38:     4        |                          |                 |  +- Ident: n
@     :     4        |                          |                 +- expr 22
@     :     4        |                          |                    +- Number: 1
@     :     4        |                          +- expr 29
@     :     4        |                             +- Call
@     :    10        |                                +- expr 25
@   35:    10        |                                |  +- Ident: fibo
                     |                                +- params
@     :     4        |                                   +- expr 28
@     :     4        |                                      +- Call
@     :     7        |                                         +- expr 40
@   15:     7        |                                         |  +- Ident: sub
                     |                                         +- params
@     :     4        |                                            +- expr 26
@   38:     4        |                                            |  +- Ident: n
@     :     4        |                                            +- expr 27
@     :     4        |                                               +- Number: 2
                     +- stmt 20
=    3:     1           +- Namespace name=std
                           +- stmt 14
=   11:     1              |  +- Annotation name=native
                           +- stmt 15
=    4:     1              |  +- Namespace name=booleans
                           |     +- stmt 28
=   28:    11              |        +- Fn name=print (native)
                           |           +- annotations
@   11:     1              |           |  +- std.native
                           |           +- params
=   39:     3              |           |  +- b: core.boolean
@     :     1              |           +- ret=[]
                           +- stmt 16
=    5:     1              |  +- Namespace name=env
                           |     +- stmt 27
=   29:    12              |        +- Fn name=args (native)
                           |           +- annotations
@   11:     1              |           |  +- std.native
                           |           +- params=[]
@     :     5              |           +- ret=std.list.List
                           +- stmt 17
=    6:     1              |  +- Namespace name=list
                           |     +- stmt 25
=    7:     5              |     |  +- Struct name=List (native)
                           |     +- stmt 26
=   30:    13              |        +- Fn name=list_get (native)
                           |           +- annotations
@   11:     1              |           |  +- std.native
                           |           +- params
=   40:     5              |           |  +- list: List
=   41:     4              |           |  +- index: number
@     :     6              |           +- ret=std.str.string
                           +- stmt 18
=    8:     1              |  +- Namespace name=numbers
                           |     +- stmt 23
=   31:    14              |     |  +- Fn name=print (native)
                           |     |     +- annotations
@   11:     1              |     |     |  +- std.native
                           |     |     +- params
=   42:     4              |     |     |  +- n: core.number
@     :     1              |     |     +- ret=[]
                           |     +- stmt 24
=   32:    15              |        +- Fn name=number_parse (native)
                           |           +- annotations
@   11:     1              |           |  +- std.native
                           |           +- params
=   43:     6              |           |  +- s: std.str.string
@     :     2              |           +- ret=core.number
                           +- stmt 19
=    9:     1                 +- Namespace name=str
                                 +- stmt 21
=   10:     6                    |  +- Struct name=string (native)
                                 +- stmt 22
=   33:    16                       +- Fn name=print (native)
                                       +- annotations
@   11:     1                          |  +- std.native
                                       +- params
=   44:     6                          |  +- s: string
@     :     1                          +- ret=[]

--- types ---
0: invalid
1: void
2: none
3: boolean
4: number
5: struct 7
6: struct 10
7: f(4, 4) : 4
8: f(4, 4) : 3
9: f(3, 3) : 3
10: f(4) : 4
11: f(3) : 1
12: f() : 5
13: f(5, 4) : 6
14: f(4) : 1
15: f(6) : 4
16: f(6) : 1
17: f() : 1
